akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala

```diff
@@ -0,0 +1,508 @@
+ * with demand of how many messages it can accept.
+ *
+ * Apart from the first message the `ProducerController` will not send more messages than requested
+ * by the `ConsumerController`.
+ *
+ * When there is demand from the consumer side the `ProducerController` sends `RequestNext` to the
+ * actual producer, which is then allowed to send one more message.
+ *
+ * Each message is wrapped by the `ProducerController` in [[ConsumerController.SequencedMessage]] with
+ * a monotonically increasing sequence number without gaps, starting at 1.
+ *
+ * In other words, the "request" protocol to the application producer and consumer is one-by-one, but
+ * between the `ProducerController` and `ConsumerController` it's window of messages in flight.
+ *
+ * The `Request` message also contains a `confirmedSeqNr` that is the acknowledgement
+ * from the consumer that it has received and processed all messages up to that sequence number.
+ *
+ * The `ConsumerController` will send [[ProducerController.Internal.Resend]] if a lost message is detected
+ * and then the `ProducerController` will resend all messages from that sequence number. The producer keeps
+ * unconfirmed messages in a buffer to be able to resend them. The buffer size is limited
+ * by the request window size.
+ *
+ * The resending is optional, and the `ConsumerController` can be started with `resendLost=false`
+ * to ignore lost messages, and then the `ProducerController` will not buffer unconfirmed messages.
+ * In that mode it provides only flow control but no reliable delivery.
+ */
+object ProducerController {
+
+  type SeqNr = Long
+
+  sealed trait InternalCommand
+
+  sealed trait Command[A] extends InternalCommand
+
+  final case class Start[A](producer: ActorRef[RequestNext[A]]) extends Command[A]
+
+  final case class RequestNext[A](
+      producerId: String,
+      currentSeqNr: SeqNr,
+      confirmedSeqNr: SeqNr,
+      sendNextTo: ActorRef[A],
+      askNextTo: ActorRef[MessageWithConfirmation[A]])
+
+  final case class RegisterConsumer[A](consumerController: ActorRef[ConsumerController.Command[A]]) extends Command[A]
+
+  /**
+   * For sending confirmation message back to the producer when the message has been confirmed.
+   * Typically used with `ask` from the producer.
+   *
+   * If `DurableProducerQueue` is used the confirmation reply is sent when the message has been
+   * successfully stored, meaning that the actual delivery to the consumer may happen later.
+   * If `DurableProducerQueue` is not used the confirmation reply is sent when the message has been
+   * fully delivered, processed, and confirmed by the consumer.
+   */
+  final case class MessageWithConfirmation[A](message: A, replyTo: ActorRef[SeqNr]) extends InternalCommand
+
+  object Internal {
+    final case class Request(confirmedSeqNr: SeqNr, upToSeqNr: SeqNr, supportResend: Boolean, viaTimeout: Boolean)
```

Comment:
would another sequence number be useful here? For the scenario that you have a fast producer and slow consumer, the consumer's buffer will full up so upTo will be say 20, then the confirmed will very slowly increase as the consumer acks the messages but in the mean time the producer will keep assuming the messages are lost and resend/extend the sequence number range. This could be avoided with a sequenceNr the consumer has stashed

Reply:
PC will not push resends. The CC will request resending when it detects gaps.
The `first` message is slightly different, it will be resent by PC.

Reply:
Ah i had a bug in the converted distributed workers sample that was resending the same messages over and over :))